contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -255,6 +263,78 @@ pub fn set_chain_config(
     }
 }
 
+/// Calculates the destination on token transfer amount.
+///
+/// The amount is calculated based on the token decimals on the source and destination chains.
+/// The calculation is done as following:
+/// 1) `destination_amount` = `source_amount` * 10 ^ (`destination_chain_decimals` - `source_chain_decimals`)
+/// 3) If new_amount is greater than the destination chain's `max_uint`, the translation
+/// fails.
+/// 4) If new_amount is zero, the translation fails.
+fn destination_amount(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    destination_chain: &ChainNameRaw,
+    token_id: TokenId,
+    source_amount: nonempty::Uint256,
+) -> Result<nonempty::Uint256, Error> {
+    let source_token = try_load_token_instance(storage, source_chain.clone(), token_id)?;
+    let destination_token = try_load_token_instance(storage, destination_chain.clone(), token_id)?;
+    let (source_decimals, destination_decimals) =
+        match (source_token.decimals, destination_token.decimals) {
+            (Some(source_decimals), Some(destination_decimals)) => {
+                (source_decimals, destination_decimals)
+            }
+            (None, None) => return Ok(source_amount),
```

Comment:
this early return here is easy to miss. I think it would be easier to read if you move this return out of the match. You could return an `Option<(u8, u8)>` from the match, and return `None` in this case. Then just check if it's `None` after the match. I think you should also check if the decimals are the same after the match as well and return early. We don't need to do the computation in that case, and I think it's easer to read.

Reply:
I don't necessarilly think the returning `Option<(u8, u8)>` makes it more readable but I do like the check to see if they are the same.